Function Calling

AdritRao opened the pull request:
<!--

This source file is part of the Stanford Biodesign Digital Health Group open-source organization

SPDX-FileCopyrightText: 2022 Stanford University and the project authors (see CONTRIBUTORS.md)

SPDX-License-Identifier: MIT

-->

# Function Calling

## :recycle: Current situation & Problem
Currently, when the user taps "Chat with all resources", all FHIR resources are sent to the LLM. This can lead to token limit errors with the API. By using function calling (https://github.com/StanfordBDHG/LLMonFHIR/issues/22), we can enable a structured interaction with health records (https://platform.openai.com/docs/guides/gpt/faq).

## :bulb: Proposed solution
The function ```get_resource_titles``` is added to the LLM. The LLM uses this function to find out which resource(s) are applicable to the user's question. Based on the function calling response, the corresponding JSON data is given to the LLM. This way, the LLM is only given FHIR resources based on the user's question. Additionally, instead of providing all resources to start the interaction, the interpreter is only fed the titles and categories of all resources to provide an overall summary.

## :gear: Release Notes 
One of the main additions to the code was the ```ChatFunctionDeclaration``` for ```get_resource_titles``` in the ```OpenAIChatView.swift```:
```
        let functions = [
            ChatFunctionDeclaration(
                name: "get_resource_titles",
                description: String(localized: "MULTIPLE_RESOURCE_FUNCTION_DESCRIPTION"),
                parameters: JSONSchema(
                    type: .object,
                    properties: [
                        "resources": .init(
                            type: .string,
                            description: String(localized: "MULTIPLE_RESOURCE_PARAMETER_DESCRIPTION"),
                            enumValues: stringResourcesArray
                        )
                    ],
                    required: ["resources"]
                )
            )
        ]
```

The chat view was also abstracted into various functions to handle function calling, parsing the JSON response from the API, and sending the LLM the relevant resources based on the function calling.

The multiple resource interpreter prompt was modified and three additional prompts were added for the function calling: ```MULTIPLE_RESOURCE_FUNCTION_DESCRIPTION```, ```MULTIPLE_RESOURCE_PARAMETER_DESCRIPTION```, and ```MULTIPLE_RESOURCE_FUNCTION_CONTEXT``` in the ```Localizable.strings``` file.

### Code of Conduct & Contributing Guidelines 

By submitting creating this pull request, you agree to follow our [Code of Conduct](https://github.com/StanfordBDHG/.github/blob/main/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/StanfordBDHG/.github/blob/main/CONTRIBUTING.md):
- [x] I agree to follow the [Code of Conduct](https://github.com/StanfordBDHG/.github/blob/main/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/StanfordBDHG/.github/blob/main/CONTRIBUTING.md).

PSchmiedmayer commented on LLMonFHIR/FHIR Display/OpenAIChatView.swift:
```diff
@@ -87,4 +77,109 @@ struct OpenAIChatView: View {
+        ]
+        
+        let query = ChatQuery(
+            model: "gpt-3.5-turbo-0613",
```
We should use the model that is configured in the OpenAI component here, the user can have the choice between different models during the onboarding flow.
We should probably only offer the function calling one here. Not sure if this requires a change in SpeziML. If that is the case, then we should create a nice issue over there to make the choices of models an option that can be configured when creating an OpenAI component.

PSchmiedmayer commented on LLMonFHIR/FHIR Display/OpenAIChatView.swift:
```diff
@@ -87,4 +77,109 @@ struct OpenAIChatView: View {
+        
+        let query = ChatQuery(
+            model: "gpt-3.5-turbo-0613",
+            messages: [
+                Chat(role: .user, content: chat.last?.content ?? ""),
+                Chat(role: .system, content: String(localized: "MULTIPLE_RESOURCE_FUNCTION_CONTEXT") + stringResourcesArray.rawValue)
+            ],
```
I foresee one problems with the current approach: We loose all the context every time we make a function call as this only contains the last message from the chat, it would be wise if we always send the complete chat when making any API request to the model. This allows it to keep the complete conversation history in mind when answering questions and e.g. only request new resources it needs access to.

Therefore I would suggest the following procedure:
1. Let's append the function call to the chat when the chat is first passed to the view (e.g. in the initializer).
2. Always use the current chat instance to make the request, this always contains the information that the function call is a possible option for the model and it can choose to invoke it at any time.

This would also dramatically reduce the different between the multiple resources chat and the current setup: The only different would be if we append the functions call message in the initializer. If we do that, then the model can use that and might at some point respond with a function call.
We can then check this in both use cases (it will only happen in the multipel resources chat one) and therefore reuse a lot of the code across the two use cases and don't need to make this a big difference path.

What do you think about that idea?

PSchmiedmayer commented on LLMonFHIR/FHIR Display/OpenAIChatView.swift:
```diff
@@ -87,4 +77,109 @@ struct OpenAIChatView: View {
+        stringResourcesArray.append("N/A")
+        let functionCallOutputArray = try await getFunctionCallOutputArray(stringResourcesArray)
+        
+        if !functionCallOutputArray.contains(where: { $0 == "N/A" }) && !functionCallOutputArray.contains(where: { $0 == " N/A" }) {
```
You can reduce this down to a statement like this, which just checks if "N/A" is somewhere in the string that you look for:
```suggestion
        if !functionCallOutputArray.contains(where: { $0.contains("N/A") }) {
```

PSchmiedmayer commented on LLMonFHIR/FHIR Display/OpenAIChatView.swift:
```diff
@@ -87,4 +77,109 @@ struct OpenAIChatView: View {
+    private func processFunctionCallOutputArray(functionCallOutputArray: [String], resourcesArray: [FHIRResource]) {
+        var stringResourcesArray = resourcesArray.map { "\($0.displayName) in \($0.resourceType)" }
+        for resource in functionCallOutputArray {
+            var stringResource = resource
+            stringResource = resource.trimmingCharacters(in: .whitespaces)
```
Can be combined in a single statement.
```suggestion
            let stringResource = resource.trimmingCharacters(in: .whitespaces)
```

PSchmiedmayer commented on LLMonFHIR/FHIR Display/OpenAIChatView.swift:
```diff
@@ -87,4 +77,109 @@ struct OpenAIChatView: View {
+            } else {
+                print("Resource '\(resource)' not found in stringResourcesArray.")
+            }
+        }
```
We will have to adapt the if-statement above (`if viewState == .idle && chat.last?.role == .user`) to also trigger the OpenAI call when the last message is a system message as appended here in the message to ensure the his reaches the OpenAI server: 
```swift
if viewState == .idle && chat.last?.role != .assistant
```

PSchmiedmayer commented on LLMonFHIR/FHIR Display/OpenAIChatView.swift:
```diff
@@ -87,4 +77,109 @@ struct OpenAIChatView: View {
+            if let index = stringResourcesArray.firstIndex(of: stringResource) {
+                let resourceJSON = resourcesArray[index].jsonDescription
+                print(resourceJSON)
+                chat.append(Chat(role: .system, content: "Based on the function get_resource_title you have requested the health record: \(stringResource). This is the associated JSON data for the resource which you will use to answer the users question: \(resourceJSON). Use this health record to answer the users question ONLY IF the health record is applicable to the question."))
```
As far as I understand the OpenAI API, we will need to add some further elements to the returned chat message to ensure that it is properly understood as a function call response:
1. The role needs to be `function`: https://github.com/MacPaw/OpenAI/blob/main/Sources/OpenAI/Public/Models/ChatQuery.swift#L11
2. The name needs to be the fiction name that the function is a response to:  https://github.com/MacPaw/OpenAI/blob/main/Sources/OpenAI/Public/Models/ChatQuery.swift#L14

PSchmiedmayer commented on LLMonFHIR/FHIR Display/OpenAIChatView.swift:
```diff
@@ -87,4 +77,109 @@ struct OpenAIChatView: View {
+
+    private func processFunctionCallOutputArray(functionCallOutputArray: [String], resourcesArray: [FHIRResource]) {
+        var stringResourcesArray = resourcesArray.map { "\($0.displayName) in \($0.resourceType)" }
+        for resource in functionCallOutputArray {
```
I would suggest that we first collect all the resources in an array, and then append them all in a single message that we append to the chat. This should the result in a single message appended to the chat, that is sub-sequentially triggering an other call to the OpenAI API as detailed in the comment at the end of this function.

PSchmiedmayer reviewed (changes requested):
Thank you for the good first draft of the function calling API integration @AdritRao! It has all the necessary pieces in place, I just had a few comments on the general structure and some OpenAI-specific features that can be improved to fully reflect the API spec and to ensure that we have a good reusability of the code across the two uses cases where we have function calls wand were we won't use them.

Let me know if you have any questions regrind the specific comments, happy to help and support in any way I can to get this PR merged 🚀

AdritRao commented:
Hi @PSchmiedmayer, thank you for your comments! I have addressed them and also added the support for function calling using the updated Spezi ML module (https://github.com/StanfordSpezi/SpeziML/pull/22). The main change after the comments was the addition of the custom ```ChatFunctionCall``` class and populating it based on the chat stream from Spezi ML.

PSchmiedmayer commented on LLMonFHIR/FHIR Display/OpenAIChatView.swift:
```diff
@@ -87,4 +77,112 @@ struct OpenAIChatView: View {
+    
+    private func processMultipleResourceChat() async throws {
+        let resourcesArray = await fhirStandard.resources
+        var stringResourcesArray = resourcesArray.map { "\($0.displayName) in \($0.resourceType)" }
```
Might be a place to reuse the code from the previous comment.

PSchmiedmayer commented on LLMonFHIR/FHIR Display/OpenAIChatView.swift:
```diff
@@ -87,4 +77,112 @@ struct OpenAIChatView: View {
+            )
+        ]
+
+        let chat = [
+            Chat(role: .user, content: chat.last?.content ?? ""),
+            Chat(role: .system, content: String(localized: "FUNCTION_CONTEXT") + stringResourcesArray.rawValue)
+        ]
```
We currently only pass in the last message here, we should pass in the complete chat here, e.g. just use the chat property to query the API.
if you want to ensure that the system message is in place, it might make more sense to add it to the chat passed in in the initializer.

PSchmiedmayer commented on LLMonFHIR/FHIR Display/OpenAIChatView.swift:
```diff
@@ -87,4 +77,112 @@ struct OpenAIChatView: View {
+        for resource in functionCallOutputArray {
+            var stringResource = resource
+            stringResource = resource.trimmingCharacters(in: .whitespaces)
+            if let index = stringResourcesArray.firstIndex(of: stringResource) {
+                let resourceJSON = resourcesArray[index].jsonDescription
```
I think it might not be wise to purely trust on the fact that the indexes are the same and they stay the same across function calls.
We should rather filter the `resourcesArray` and try to get the resource, e.g., using the extension that I recommended in the other comment.

PSchmiedmayer commented on LLMonFHIR/FHIR Display/OpenAIChatView.swift:
```diff
@@ -87,4 +77,112 @@ struct OpenAIChatView: View {
+                This is the associated JSON data for the resources which you will use to answer the users question: \(resourceJSON).
+                Use this health record to answer the users question ONLY IF the health record is applicable to the question.
+                """
+                chat.append(Chat(role: .function, content: functionContent, name: "get_resource_titles"))
```
I think that the response should rather be a `.system` message if I correctly understood this from the documentation. 
Appending multiple responses sounds good to me 👍

AdritRao commented:
Hi @PSchmiedmayer, thank you for your feedback! I have addressed the comments in the latest commit. I modified the code so that the ```.system``` message with function context is provided at the beginning of the chat. The function call is now performed with the full chat as input. I have done some testing and so far, the LLM is able to prevent itself from re-querying the same records!

PSchmiedmayer reviewed (approved):
Thank you for the improvements @AdritRao

I took a look at the PR and made some smaller refinements and code improvements that should not hold us up from merging the PR with doing an other round of feedback 👍

You can see the diff of my changes here: https://github.com/StanfordBDHG/LLMonFHIR/pull/29/files/a69006baa6632549441d85e94229351f0b41ccf8..bc86ecbe757ad8995c0aa8a6a0e1a0770102aec5 ... the biggest diff comes from reducing the indent a bit by using guard statements and removing code where Xcode warned us that types and properties were unused.

I have approved the PR in its current state and you can merge it at any time that works for you  🎉

Thank you for all the work and effort that went into the PR and improving LLMonFHIR!
